crunch-core/src/main/java/org/apache/crunch/util/CrunchRenameCopyListing.java

```diff
@@ -0,0 +1,261 @@
+
+    pathRenames = new HashMap<>();
+
+    String[] pathRenameConf = configuration.getStrings(DISTCP_PATH_RENAMES);
+    if (pathRenameConf == null) {
+      throw new IllegalArgumentException("Missing required configuration: " + DISTCP_PATH_RENAMES);
+    }
+    for (String pathRename : pathRenameConf) {
+      String[] pathRenameParts = pathRename.split(":");
+      if (pathRenameParts.length != 2) {
+        throw new IllegalArgumentException("Invalid path rename format: " + pathRename);
+      }
+      if (pathRenames.put(pathRenameParts[0], pathRenameParts[1]) != null) {
+        throw new IllegalArgumentException("Invalid duplicate path rename: " + pathRenameParts[0]);
+      }
+    }
+    LOG.info("Loaded {} path rename entries", pathRenames.size());
+  }
+
+  @Override
+  public void doBuildListing(SequenceFile.Writer fileListWriter, DistCpOptions options) throws IOException {
+    try {
+      for (Path path : options.getSourcePaths()) {
+        FileSystem sourceFS = path.getFileSystem(getConf());
+        final boolean preserveAcls = options.shouldPreserve(FileAttribute.ACL);
+        final boolean preserveXAttrs = options.shouldPreserve(FileAttribute.XATTR);
+        final boolean preserveRawXAttrs = options.shouldPreserveRawXattrs();
+        path = makeQualified(path);
+
+        FileStatus rootStatus = sourceFS.getFileStatus(path);
+        Path sourcePathRoot = computeSourceRootPath(rootStatus, options);
+
+        FileStatus[] sourceFiles = sourceFS.listStatus(path);
+        boolean explore = (sourceFiles != null && sourceFiles.length > 0);
+        if (!explore || rootStatus.isDirectory()) {
+          CopyListingFileStatus rootCopyListingStatus = DistCpUtils.toCopyListingFileStatus(sourceFS, rootStatus, preserveAcls,
+              preserveXAttrs, preserveRawXAttrs);
+          writeToFileListingRoot(fileListWriter, rootCopyListingStatus, sourcePathRoot, options);
+        }
+        if (explore) {
+          for (FileStatus sourceStatus : sourceFiles) {
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Recording source-path: " + sourceStatus.getPath() + " for copy.");
+            }
+            CopyListingFileStatus sourceCopyListingStatus = DistCpUtils.toCopyListingFileStatus(sourceFS, sourceStatus,
+                preserveAcls && sourceStatus.isDirectory(), preserveXAttrs && sourceStatus.isDirectory(),
+                preserveRawXAttrs && sourceStatus.isDirectory());
+            writeToFileListing(fileListWriter, sourceCopyListingStatus, sourcePathRoot, options);
+
+            if (isDirectoryAndNotEmpty(sourceFS, sourceStatus)) {
+              if (LOG.isDebugEnabled()) {
+                LOG.debug("Traversing non-empty source dir: " + sourceStatus.getPath());
+              }
+              traverseNonEmptyDirectory(fileListWriter, sourceStatus, sourcePathRoot, options);
+            }
+          }
+        }
+      }
+      fileListWriter.close();
+      fileListWriter = null;
```

Comment:
won't this potentially cause an NPE in the finally block as this reference is being set equal to null?

Reply:
Yes, that's a bug, thanks for catching it. I will fix this.